File: tests/workflow/producer/rust/src/producer/responses/groupb_groupc_structb.rs
use super::{identification, producer::Control, protocol, Context};
use crate::stat::Alias;
use crate::test::samples;
use clibri::server;
use uuid::Uuid;

type BroadcastStructD = (Vec<Uuid>, protocol::StructD);
type BroadcastStructF = (Vec<Uuid>, protocol::StructF);
type BroadcastStructJ = (Vec<Uuid>, protocol::StructJ);

pub enum Response {
    CaseB((protocol::StructB, BroadcastStructD, BroadcastStructF)),
    CaseC(protocol::StructC),
    CaseD((protocol::StructD, BroadcastStructJ)),
}

#[allow(unused_variables)]
pub async fn response<'c, E: server::Error, C: server::Control<E> + Send + Clone>(
    identification: &identification::Identification,
    filter: &identification::Filter<'_>,
    context: &mut Context,
    request: &protocol::GroupB::GroupC::StructB,
    control: &Control<E, C>,
) -> Result<Response, protocol::GroupB::GroupC::StructA> {
    let index = context.get_index(identification.uuid(), Alias::GroupBGroupCStructB);
    if index == 1 {
        context.inc_stat(identification.uuid(), Alias::StructB);
        context.inc_stat(identification.uuid(), Alias::StructD);
        context.inc_stat(identification.uuid(), Alias::StructF);
        Ok(Response::CaseB((
            samples::struct_b::get(),
            (vec![identification.uuid()], samples::struct_d::get()),
            (vec![identification.uuid()], samples::struct_f::get()),
        )))
    } else if index == 2 {
        context.inc_stat(identification.uuid(), Alias::StructC);
        Ok(Response::CaseC(samples::struct_c::get()))
    } else if index == 3 {
        context.inc_stat(identification.uuid(), Alias::StructD);
        context.inc_stat(identification.uuid(), Alias::StructJ);
        Ok(Response::CaseD((
            samples::struct_d::get(),
            (vec![identification.uuid()], samples::struct_j::get()),
        )))
    } else {
        context.inc_stat(identification.uuid(), Alias::GroupBGroupCStructA);
        Err(samples::group_b::group_c::struct_a::get())
    }
}

File: tests/workflow/producer/typescript/src/producer/responses/groupb.groupc.structb.ts
import { Response } from "../implementation/responses/groupb.groupc.structb";
import {
	Context,
	Producer,
	Identification,
	Filter,
	Protocol,
} from "../implementation/responses";
import { Alias } from "../../stat";
import { panic } from "../../tools";
import { Scope } from "../implementation/scope";

export function response(
	request: Protocol.GroupB.GroupC.StructB,
	scope: Scope
): Promise<Response> {
	const stat = scope.context.getStat(scope.consumer.uuid());
	stat.case(Alias.GroupBGroupCStructB);
	const index = stat.index(Alias.GroupBGroupCStructB);
	if (index === 0) {
		return Promise.resolve(
			new Response(Protocol.StructB.defaults())
				.broadcast([scope.consumer.uuid()])
				.StructD(Protocol.StructD.defaults())
				.broadcast([scope.consumer.uuid()])
				.StructF(Protocol.StructF.defaults())
		);
	} else if (index === 1) {
		return Promise.resolve(new Response(Protocol.StructC.defaults()));
	} else if (index === 2) {
		stat.case(Alias.StructJ);
		return Promise.resolve(
			new Response(Protocol.StructD.defaults())
				.broadcast([scope.consumer.uuid()])
				.StructJ(Protocol.StructJ.defaults())
		);
	} else if (index === 3) {
		return Promise.resolve(
			new Response(Protocol.GroupB.GroupC.StructA.defaults())
		);
	} else {
		panic(
			`Unexpected index for Protocol.GroupB.GroupC.StructB request: ${index}`
		);
		return Promise.reject(
			new Error(
				`Handler for Protocol.GroupB.GroupC.StructB isn't implemented.`
			)
		);
	}
}
